src/storage/src/sink/kafka.rs

```diff
@@ -1300,7 +1317,13 @@ where
             // Only ever emit progress records if this operator/worker received
             // updates. Only on worker receives all the updates and we don't want
             // the other workers to also emit END records.
-            if is_active_worker {
+            //
+            // NOTE: Very subtle, but make sure we have not been shut down in
+            // the mean time. We only want to downgrade/emit progress while
+            // we're "still running". This flag can change from where we check
+            // it all the way at the beginning of the block because of the await
+            // points.
+            if is_active_worker && !s.shutdown_flag.load(Ordering::SeqCst) {
```

Comment:
Is this one a correctness issue, or just liveness? (Before this change, are we at risk of having committed a partial batch?)

Reply:
Correctness, because we might allow compaction (of the input collection) to sth that we don't want. Side note: it's a bit confusing that `maybe_emit_progress` also updates the shared frontier tracking that is, in the end, used to allow compaction on the input collection. That's on me!

Reply:
can you explain this subtlety a bit more? when do we expect to set the shutdown flag without returning? (which would imply that before we continue processing, we'd check the flag again)

Reply:
Huh. Can you elaborate? I can't find a code path that will update the frontier tracking without committing a transaction first. Which implies that even if we're shutting down the data should already be safely in Kafka.

Reply:
a key insight from mine and aljoscha's chat (that I'm writing down now in case I come back and look at this in the future) is that I had forgotten that we also pass off the shutdown flag in the `KafkaSinkToken` -- and if that token is dropped, we set the shutdown flag.  The issue here is that we shouldn't read the frontier unless the shutdown flag is unset

Reply:
okay I looked into this more (and we chatted offline) but for the record:  this isn't actually necessary here.  The `async_op` macro caches a value of the input frontier and passes it through per invocation of this function.  That means the `frontiers` (and thus `frontier`) cannot change in a single invocation of the function.  The check at the top of the function should be sufficient for correctness here.